web/modules/cm-espells/src/espells/interface.ts

```diff
@@ -0,0 +1,128 @@
+import { AbstractWorkerBase, Comlink } from "@wikijump/comlink"
+import Locale from "@wikijump/fluent"
+import { dedupe, LazySingleton, Pref } from "@wikijump/util"
+import type { Espells } from "espells"
+import DICTIONARIES from "../dicts"
+import type { FlaggedWord, Word } from "../types"
+import RemoteWorker from "./worker?worker"
+
+const RemoteClassSingleton = new LazySingleton(() =>
+  Comlink.wrap<typeof Espells>(new RemoteWorker())
+)
+
+export class EspellsWorker extends AbstractWorkerBase.of<Espells>([
+  "add",
+  "addDictionary",
+  "data",
+  "lookup",
+  "remove",
+  "stems",
+  "suggest"
+]) {
+  declare locale: string
+
+  constructor(locale = "en") {
+    super()
+    this.set(locale)
+  }
+
+  async set(locale: string) {
+    this.locale = localeLanguage(locale)
+    if (this.loaded) await this.start(true)
+  }
+
+  protected async createWorker() {
+    if (DICTIONARIES.hasOwnProperty(this.locale)) {
+      const { aff: affURL, dic: dicURL } = await DICTIONARIES[this.locale]()
+
+      // we can't use the static fromURL method on Espells because of Comlink limitations
+      // so it has to be reimplemented basically - we have to resolve the urls
+
+      const aff = await fetch(affURL).then(res => res.text())
+      const dic = Array.isArray(dicURL)
+        ? await Promise.all(dicURL.map(dic => fetch(dic))).then(res =>
```

Comment:
This is a bit difficult to read, maybe worth adding a comment or two explaining what the parts of the callbacks are doing?

Reply:
Yeah, or I should maybe rewrite it. It's just trying to be fast by doing all the fetch requests in parallel.